fix(types): Allow no relative url to fetch RESTful api list

krystofwoldrich opened the pull request:
fix(get): Allow no relative url to fetch RESTful api list

<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#Pull-Request
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

I've noticed examples in the `README.md`, where the `get` method was used without any parameters. But the TS typing didn't allowed such a calls. In my opinion the call without any arguments looks cleaner than call with empty string to get list from a resource.

posva commented on src/index.ts:
```diff
@@ -66,9 +66,15 @@ export interface MandeInstance {
-  get(url: string | number, options: Options<'text'>): Promise<string>
-  get<T = unknown>(url: string | number, options?: Options): Promise<T>
+  get(
+    url: string | number | undefined,
```
this is not correct: it shouldn't allow `undefined`

posva commented on src/index.ts:
```diff
@@ -66,9 +66,15 @@ export interface MandeInstance {
+    url: string | number | undefined,
+    options: Options<'text'>
+  ): Promise<string>
+  get<T = unknown>(url?: string | number, options?: Options): Promise<T>
```
there should also be a version that allows `get(options?: Options)`. This will also require adding some types to `delete()` BTW

posva commented on src/index.ts:
```diff
@@ -343,7 +349,7 @@ export function mande(
 
     // these two have no body
-    get: (url: string, options?: Options) => _fetch('GET', url, null, options),
+    get: (url?: string, options?: Options) => _fetch('GET', url, null, options),
```
these will also need to get updated to `urlOrOptions?: string | number | Options`

posva commented on __tests__/index.spec.ts:
```diff
@@ -41,6 +41,13 @@ describe('mande', () => {
     expect(fetchMock).toHaveFetched('/api/2')
   })
 
+  it('can use get without parameters', async () => {
```
Can you also add a test that you can pass options to `api.get({ ... })`?

krystofwoldrich replied:
Sure thing, I’ll add it.

posva reviewed (changes requested):
Thanks!

krystofwoldrich commented on docs/mande.mandeinstance.get.md:
```diff
@@ -16,7 +16,7 @@ get(url: string | number, options: Options<'response'>): Promise<Response>;
 |  Parameter | Type | Description |
 |  --- | --- | --- |
-|  url | string \| number | relative url to send the request to |
+|  url | string \| number \| optional | relative url to send the request to |
```
OK, I didn’t know that, how do I generate it?

posva replied:
It will be automatic

krystofwoldrich commented:
Okay. I've changed the typing and added a bit of logic to correctly recognise url and options. Also I've added 2 more tests. 

Let me know, what do you think.

krystofwoldrich commented:
Any updates?

posva commented on src/index.ts:
```diff
@@ -343,9 +345,22 @@ export function mande(
-    get: (url: string, options?: Options) => _fetch('GET', url, null, options),
-    delete: (url: string, options?: Options) =>
-      _fetch('DELETE', url, null, options),
+    get: (
```
rather than updating these, I think this should doable only changing the types of `_fetch`

krystofwoldrich replied:
okay, I'll take a look.

posva commented on src/index.ts:
```diff
@@ -63,11 +63,12 @@ export interface MandeInstance {
    */
   get(url: string | number, options: Options<'response'>): Promise<Response>
   get(url: string | number, options: Options<'text'>): Promise<string>
+  get<T = unknown>(options?: Options): Promise<T>
```
this probably also needs the other overrides as well and some test types

krystofwoldrich replied:
You mean `Options<'response'>` and `Options<'text'>`.

And test types, I don't follow?

krystofwoldrich commented:
@posva Thanks for the tips. I've change it. I hope it's okay, you can take a look.

posva commented:
Thank you, I added a few last changes and merged it in #379
